File: src/algorithms/sets/operations/set-equality/__tests__/set-equality.test.ts
import { describe, it, expect } from "vitest";
import { setEquality } from "../sources/set-equality.ts?fn";

describe("setEquality", () => {
  it("returns true for the default input (same elements, different order)", () => {
    expect(setEquality([3, 1, 2], [2, 3, 1])).toEqual({ isEqual: true });
  });

  it("returns true for identical arrays", () => {
    expect(setEquality([1, 2, 3], [1, 2, 3])).toEqual({ isEqual: true });
  });

  it("returns false when B has an element not in A", () => {
    expect(setEquality([1, 2, 3], [1, 2, 9])).toEqual({ isEqual: false });
  });

  it("returns false when A has more unique elements than B", () => {
    expect(setEquality([1, 2, 3, 4], [1, 2, 3])).toEqual({ isEqual: false });
  });

  it("returns false when B has more unique elements than A", () => {
    expect(setEquality([1, 2, 3], [1, 2, 3, 4])).toEqual({ isEqual: false });
  });

  it("returns true for two empty arrays", () => {
    expect(setEquality([], [])).toEqual({ isEqual: true });
  });

  it("returns false when A is empty and B is non-empty", () => {
    expect(setEquality([], [1])).toEqual({ isEqual: false });
  });

  it("returns false when B is empty and A is non-empty", () => {
    expect(setEquality([1], [])).toEqual({ isEqual: false });
  });

  it("handles duplicates in A — unique count matters", () => {
    // A = {1, 2}, B = {1, 2, 3} — not equal
    expect(setEquality([1, 1, 2], [1, 2, 3])).toEqual({ isEqual: false });
  });

  it("handles duplicates in B — unique count matters", () => {
    // A = {1, 2, 3}, B = {1, 2} — not equal
    expect(setEquality([1, 2, 3], [1, 1, 2])).toEqual({ isEqual: false });
  });

  it("returns true when both arrays have duplicates but same unique set", () => {
    // A = {1, 2, 3}, B = {1, 2, 3}
    expect(setEquality([1, 1, 2, 3], [1, 2, 2, 3])).toEqual({ isEqual: true });
  });

  it("handles single-element equal arrays", () => {
    expect(setEquality([7], [7])).toEqual({ isEqual: true });
  });

  it("handles single-element unequal arrays", () => {
    expect(setEquality([7], [8])).toEqual({ isEqual: false });
  });
});

File: src/algorithms/sets/operations/set-equality/SetEquality_test.java
public class SetEquality_test {

    public static void main(String[] args) {
        // same elements different order
        assert SetEquality.setEquality(new int[]{3, 1, 2}, new int[]{2, 3, 1}) == true;

        // identical arrays
        assert SetEquality.setEquality(new int[]{1, 2, 3}, new int[]{1, 2, 3}) == true;

        // B has element not in A
        assert SetEquality.setEquality(new int[]{1, 2, 3}, new int[]{1, 2, 9}) == false;

        // A has more unique elements
        assert SetEquality.setEquality(new int[]{1, 2, 3, 4}, new int[]{1, 2, 3}) == false;

        // B has more unique elements
        assert SetEquality.setEquality(new int[]{1, 2, 3}, new int[]{1, 2, 3, 4}) == false;

        // both empty
        assert SetEquality.setEquality(new int[]{}, new int[]{}) == true;

        // A empty B non-empty
        assert SetEquality.setEquality(new int[]{}, new int[]{1}) == false;

        // B empty A non-empty
        assert SetEquality.setEquality(new int[]{1}, new int[]{}) == false;

        // duplicates same unique set
        assert SetEquality.setEquality(new int[]{1, 1, 2, 3}, new int[]{1, 2, 2, 3}) == true;

        // single element equal
        assert SetEquality.setEquality(new int[]{7}, new int[]{7}) == true;

        // single element not equal
        assert SetEquality.setEquality(new int[]{7}, new int[]{8}) == false;

        System.out.println("All tests passed!");
    }
}
